package/MDAnalysis/lib/formats/xdrlib.pyx

```diff
@@ -0,0 +1,662 @@
+            If you seek for more frames then are available
+        """
+        cdef int offset
+        if frame == 0:
+            offset = 0
+        else:
+            if frame >= self.offsets.size:
+                raise RuntimeError('Trying to seek over max number of frames')
+            offset = self.offsets[frame]
+        self.reached_eof = False
+        ok = xdr_seek(self.xfp, offset, SEEK_SET)
+        if ok != 0:
+            raise RuntimeError("XDR seek failed, error-message={}".format(
+                error_message[ok]))
+        self.current_frame = frame
+
+    @property
+    def offsets(self):
+        """get byte offsets from current xdr file"""
+        if not self._has_offsets:
+            self._offsets = self.calc_offsets()
+            self._has_offsets = True
+        return self._offsets
+
+    def set_offsets(self, offsets):
+        self._offsets = offsets
+        self._has_offsets = True
+
+    def tell(self):
+        """Get current frame"""
+        return self.current_frame
+
+
+TRRFrame = namedtuple('TRRFrame', 'x v f box step time lmbda hasx hasv hasf')
+
+
+cdef class TRRFile(_XDRFile):
+    """File-like wrapper for gromacs TRR files.
+
+    This class can be similar to the normal file objects in python. The read()
+    function will return a frame and all information in it instead of a single
+    line. Additionally the context-manager protocoll is supported as well.
+
+    Parameters
+    ----------
+    fname: str
+        The filename to open.
+    mode: ('r', 'w')
+        The mode in which to open the file, either 'r' read or 'w' write
+
+    Examples
+    --------
+    >>> from MDAnalysis.lib.formats.xdrlib import TRRFile
+    >>> with TRRFile('foo.trr') as f:
+    >>>     for frame in f:
+    >>>         print(frame.x)
+    """
+
+    def _calc_natoms(self, fname):
+        cdef int n_atoms
+        return_code = read_trr_natoms(fname, &n_atoms)
+        return return_code, n_atoms
+
+
+    def calc_offsets(self):
+        """read byte offsets from TRR file directly"""
+        if not self.is_open:
+            return np.array([])
+        cdef int n_frames = 0
+        cdef int est_nframes = 0
+        cdef int64_t* offsets = NULL
+        ok = read_trr_n_frames(self.fname, &n_frames, &est_nframes, &offsets);
+        if ok != 0:
+            raise RuntimeError("TRR couldn't calculate offsets, error={}".format(
+                error_message[ok]))
+        # the read_xtc_n_frames allocates memory for the offsets with an
+        # overestimation. This number is saved in est_nframes and we need to
+        # tell the new numpy array about the whole allocated memory to avoid
+        # memory leaks.
+        cdef np.ndarray dims = np.array([est_nframes], dtype=np.int64)
+        # this handles freeing the allocated memory correctly.
+        cdef np.ndarray nd_offsets = ptr_to_ndarray(<void*> offsets, dims, np.NPY_INT64)
+        return nd_offsets[:n_frames]
+
+    def read(self):
```

Comment:
i thought of that as well. But since I can't detect any performance regressions right now I thought this could be for another PR later that concentrates on performance issues alone. I wanted to get correct behavior in this one.